chia-protocol/src/bytes.rs

```diff
@@ -211,153 +253,87 @@ impl<N, const LEN: usize> FromClvm<N> for BytesImpl<LEN> {
+    type Error = TryFromSliceError;
 
-impl<const N: usize> From<&BytesImpl<N>> for [u8; N] {
-    fn from(v: &BytesImpl<N>) -> [u8; N] {
-        v.0
+    fn try_from(value: &[u8]) -> Result<Self, TryFromSliceError> {
+        Ok(Self(value.try_into()?))
     }
 }
 
-impl<'a, const N: usize> From<&'a BytesImpl<N>> for &'a [u8] {
-    fn from(v: &'a BytesImpl<N>) -> &'a [u8] {
-        &v.0
-    }
-}
+impl<const N: usize> TryFrom<Vec<u8>> for BytesImpl<N> {
+    type Error = TryFromSliceError;
 
-impl<const N: usize> BytesImpl<N> {
-    pub fn to_vec(&self) -> Vec<u8> {
-        self.0.to_vec()
+    fn try_from(value: Vec<u8>) -> Result<Self, TryFromSliceError> {
+        value.as_slice().try_into()
     }
 }
 
-impl<const N: usize> AsRef<[u8]> for BytesImpl<N> {
-    fn as_ref(&self) -> &[u8] {
-        &self.0
-    }
-}
-impl<const N: usize> Deref for BytesImpl<N> {
-    type Target = [u8];
-    fn deref(&self) -> &[u8] {
-        &self.0
-    }
-}
-impl<const N: usize> Debug for BytesImpl<N> {
-    fn fmt(&self, formatter: &mut Formatter<'_>) -> std::result::Result<(), std::fmt::Error> {
-        formatter.write_str(&hex::encode(self.0))
+impl<const N: usize> TryFrom<&Vec<u8>> for BytesImpl<N> {
+    type Error = TryFromSliceError;
+
+    fn try_from(value: &Vec<u8>) -> Result<Self, TryFromSliceError> {
+        value.as_slice().try_into()
     }
 }
-impl<const N: usize> fmt::Display for BytesImpl<N> {
-    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
-        formatter.write_str(&hex::encode(self.0))
+
+impl<const N: usize> From<BytesImpl<N>> for Vec<u8> {
+    fn from(value: BytesImpl<N>) -> Self {
+        value.to_vec()
     }
 }
 
-impl<const N: usize> PartialEq<&[u8]> for BytesImpl<N> {
-    fn eq(&self, lhs: &&[u8]) -> bool {
-        self.0 == *lhs
+impl<const N: usize> From<[u8; N]> for BytesImpl<N> {
+    fn from(value: [u8; N]) -> Self {
+        Self(value)
     }
 }
 
-impl<const N: usize> PartialEq<BytesImpl<N>> for &[u8] {
-    fn eq(&self, lhs: &BytesImpl<N>) -> bool {
-        self == &lhs.0
+impl<const N: usize> From<&[u8; N]> for BytesImpl<N> {
+    fn from(value: &[u8; N]) -> Self {
+        Self(*value)
     }
 }
 
-impl<const N: usize> PartialEq<&[u8; N]> for BytesImpl<N> {
-    fn eq(&self, lhs: &&[u8; N]) -> bool {
-        &self.0 == *lhs
+impl<const N: usize> From<BytesImpl<N>> for [u8; N] {
+    fn from(value: BytesImpl<N>) -> Self {
+        value.0
     }
 }
 
-impl<const N: usize> PartialEq<BytesImpl<N>> for &[u8; N] {
-    fn eq(&self, lhs: &BytesImpl<N>) -> bool {
-        *self == &lhs.0
+impl<'a, const N: usize> From<&'a BytesImpl<N>> for &'a [u8; N] {
+    fn from(value: &'a BytesImpl<N>) -> &'a [u8; N] {
+        &value.0
     }
 }
 
-impl<const N: usize> PartialEq<[u8; N]> for BytesImpl<N> {
-    fn eq(&self, lhs: &[u8; N]) -> bool {
-        &self.0 == lhs
+impl<const N: usize> From<&BytesImpl<N>> for [u8; N] {
+    fn from(value: &BytesImpl<N>) -> [u8; N] {
+        value.0
```

Comment:
how does this work? This is moving the array out of `value`, how come it's allowed to be taken as an immutable reference?
If it's making a copy, how come you don't need to say `.clone()`?

Reply:
Both `BytesImpl<N>` and `[u8; N]` implement the `Copy` trait, so they can be freely copied without worrying about ownership or references. It doesn't actually modify the original byte array, just copies the bits directly.

https://doc.rust-lang.org/std/marker/trait.Copy.html